### DISCUSSION TITLE
Writing back data using soa_to_aos2 (and its siblings)

### DISCUSSION THREAD

**Discussion by CiderMan** [ispc/ispc#2159] [Q&A] [answered]
We are using ISPC to optimize key algorithms, often operating on complex values. As most of the code is written in C++, the complex data is stored in AOS form. To illustrate:
```C++
struct tComplex {float real, imag;};
void do_stuff(uniform tComplex data[], uniform int count)
{
     foreach (i = 0...count)
     {
         tComplex value = data[i];
         value.real += 1;
         value.imag -= 1;
         data[i] = value;
     }
}
```
As [you can see on Godbolt](https://ispc.godbolt.org/z/rPY8Pnqjz), this gives scatter/gather warnings:
```
<source>:6:27: Performance Warning: Gather required to load value. 
         tComplex value = data[i];
                          ^^^^^^^

<source>:9:10: Performance Warning: Scatter required to store value. 
         data[i] = value;
         ^^^^^^^
```
Previously, we've hand coded various approaches, depending on the algorithm. However, with the addition of aos_to_soa2, we can potentially do something like:
```C++
struct tComplex {float real, imag;};
void do_stuff(uniform tComplex data[], uniform int count)
{
    uniform int base = 0;
     foreach (i = 0...count)
     {
         float real;
         float imag;
         aos_to_soa2((float * uniform)&data[base], &real, &imag);
         real += 1;
         imag -= 1;
         soa_to_aos2(real, imag, (float * uniform)&data[base]);
         base += programCount;
     }
}
```
This fixes the performance warnings but leaves me with a question: where `count` is not an exact multiple of `programCount`, will this result in read and write overruns at the end of the `data`? That is, do these functions always read and write a full 2x the gang size or, for the last iteration of the `foreach` loop, does it only read and write the 'valid' data?

**Comment by dbabokin** [comment, marked as answer]
`aos_to_soa*` functions are not taking into account the mask, so if `count` is not a multiple of `programCount`, then there will be reads and write past the last value in the array.

We should probably specifically mention that in the documentation.

As for workarounds for this problem, I see a couple way of dealing with it (without fixing the compiler):
- ensure that data buffers have allocation size multiple of `programCount`, so reading/writing past count will not be a problem.
- trim `count` to multiple of `programCount` and have manual "reminder" dealing with reaming iterations.

Also, I assume that the example is not a real code and in the real one you have more operations over the data between `aos_to_soa` and `soa_to_aos`. As in the example above it's easier to do `+/-1` operations without doing data transformations - it will not be very elegant, but more efficient.

**Reply by CiderMan** [reply]
Thanks Dmitry. I agree that, given the scope for buffer overflow, it would be good to mention this in the documentation. I'll take a look at creating a PR with this clarification.

(and, yes, the example is not real code, as you suspected)
